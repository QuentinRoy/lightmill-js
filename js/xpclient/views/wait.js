/*jslint nomen: true, browser:true*/
/*global define */

define(['spin', 'jquery'], function (Spinner, $) {
    "use strict";

    function WaitView(parentView) {
        this._parentView = parentView;
        this._spinView = $('<div class="full-parent"></div>');
        this._spinner = new Spinner(this._opts);
    }

    WaitView.prototype = {

        _opts: {
            lines: 11, // The number of lines to draw
            length: 20, // The length of each line
            width: 10, // The line thickness
            radius: 30, // The radius of the inner circle
            corners: 1, // Corner roundness (0..1)
            rotate: 0, // The rotation offset
            direction: 1, // 1: clockwise, -1: counterclockwise
            color: '#5C5C5C', // #rgb or #rrggbb or array of colors
            speed: 1, // Rounds per second
            trail: 60, // Afterglow percentage
            shadow: false, // Whether to render a shadow
            hwaccel: false, // Whether to use hardware acceleration
            className: 'spinner', // The CSS class to assign to the spinner
            zIndex: 2e9, // The z-index (defaults to 2000000000)
            top: 'auto', // Top position relative to parent in px
            left: 'auto' // Left position relative to parent in px
        },

        startWaiting: function () {
            this._spinView.appendTo(this._parentView);
            this._spinner.spin(this._spinView[0]);
        },
        stopWaiting: function () {
            this._spinner.stop();
            this._spinView.remove();
        }
    };

    return WaitView;

});