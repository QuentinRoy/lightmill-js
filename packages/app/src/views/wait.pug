.lightmill-base-app.lightmill-base-app-wait
  .lightmill-puck-wrapper
    .lightmill-puck.lightmill-puck-1
    .lightmill-puck.lightmill-puck-2
