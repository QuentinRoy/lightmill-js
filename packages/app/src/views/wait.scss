@import './common';

$puck-wrapper-size: 60px;
$puck-radius: 5px;
$puck-color: #5e5e5e;
$background-color: #d1d1d1;
$rotation-duration: 1s;
$animation-name: #{$app-class}-wait-spin;

.#{$app-class}-wait {
  background-color: $background-color;

  .#{$class-prefix}puck-wrapper {
    position: absolute;
    top: 50%;
    left: 50%;
    animation: $animation-name $rotation-duration infinite linear;
    margin-top: -$puck-wrapper-size / 2;
    margin-left: -$puck-wrapper-size / 2;
    width: $puck-wrapper-size;
    height: $puck-wrapper-size;

    .#{$class-prefix}puck {
      position: absolute;
      border-radius: $puck-radius;
      background-color: $puck-color;
      width: $puck-radius * 2;
      height: $puck-radius * 2;
    }

    .#{$class-prefix}puck-1 {
      top: 0;
      left: 0;
    }

    .#{$class-prefix}puck-2 {
      right: 0;
      bottom: 0;
    }
  }
}

@keyframes #{$animation-name} {
  0%   {transform: rotate(0deg)}
  100% {transform: rotate(360deg)}
}
