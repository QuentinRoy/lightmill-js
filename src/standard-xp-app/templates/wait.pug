.xp-app-base.xp-app-base-wait
  .puck-wrapper
    .puck
    
