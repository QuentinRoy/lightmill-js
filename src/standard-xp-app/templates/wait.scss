@import './common';

$puck-wrapper-size: 400px;
$puck-radius: 25px;
$puck-color: #5e5e5e;
$background-color: #d1d1d1;
$rotation-duration: 1s;
$animation-name: #{$app-class}-wait-spin;

.#{$app-class}-wait {
  background-color: $background-color;

  .puck-wrapper {
    position: absolute;
    top: 50%;
    left: 50%;
    margin-top: -$puck-wrapper-size / 2;
    margin-left: -$puck-wrapper-size / 2;
    width: $puck-wrapper-size;
    height: $puck-wrapper-size;

    .puck {
      position: absolute;
      top: 0;
      left: 0;
      animation: $animation-name $rotation-duration infinite linear;
      border-radius: $puck-radius;
      background-color: $puck-color;
      width: $puck-radius * 2;
      height: $puck-radius * 2;
    }
  }
}

@keyframes #{$animation-name} {
  0%   {transform: rotate(0deg)}
  100% {transform: rotate(360deg)}
}
