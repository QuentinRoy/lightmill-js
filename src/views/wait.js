import Spinner from 'spin';

const spinnerArgs = {
  lines: 11, // The number of lines to draw
  length: 20, // The length of each line
  width: 10, // The line thickness
  radius: 30, // The radius of the inner circle
  corners: 1, // Corner roundness (0..1)
  rotate: 0, // The rotation offset
  direction: 1, // 1: clockwise, -1: counterclockwise
  color: 'white', // #rgb or #rrggbb or array of colors
  speed: 1, // Rounds per second
  trail: 60, // Afterglow percentage
  shadow: false, // Whether to render a shadow
  hwaccel: false, // Whether to use hardware acceleration
  className: 'spinner', // The CSS class to assign to the spinner
  zIndex: 2e9, // The z-index (defaults to 2000000000)
  top: 'auto', // Top position relative to parent in px
  left: 'auto' // Left position relative to parent in px
};

export default class WaitView {
  constructor(parentNode) {
    this._parentNode = parentNode;
    this._spinDiv = document.createElement('div');
    Object.assign(this._spinDiv.style, {
      position: 'absolute',
      width: '100%',
      height: '100%',
      margin: 0,
      padding: 0,
      top: 0,
      left: 0
    });
    this._spinDiv.style.backgroundColor = 'black';
    this._spinner = new Spinner(spinnerArgs);
    this._isShown = false;
  }
  show() {
    if (!this.isShown) {
      this._parentNode.appendChild(this._spinDiv);
      this._spinner.spin(this._spinDiv);
      this._isShown = true;
    }
  }
  hide() {
    if (this.isShown) {
      this._spinner.stop();
      if (this._parentNode.contains(this._spinDiv)) {
        this._parentNode.removeChild(this._spinDiv);
      }
      this._isShown = false;
    }
  }
  get isShown() {
    return this._isShown;
  }
}
