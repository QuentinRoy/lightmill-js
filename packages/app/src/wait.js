/* eslint-disable no-param-reassign */

import waitTemplate from './views/wait.pug';
import './views/wait.scss';

export default (node, message) => {
  node.innerHTML = waitTemplate({ message });

  return {
    remove() {
      node.innerHTML = '';
    }
  };
};
