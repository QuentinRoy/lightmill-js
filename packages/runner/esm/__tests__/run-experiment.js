import wait from 'wait-then';
import runTrials from '../run-trials';
import runExperiment from '../run-experiment';

jest.mock('../run-trials');

// Create a group of delayed spies, each spies create from this group is given
// a type and all spy calls can be monitored (in order) with the `all` spy.
// Also, all spies return a promise that resolves asynchronously. The resolution
// of the promise can be monitored as well (calls all with the prefix *).
const createDelayedSpiesGroup = () => {
  // Create the main spy, registering all spies call and spies return promise
  // resolutions.
  const all = jest.fn().mockName('spyGroup#all');
  // A function to add a new spy to the spy group.
  const createSpy = (
    type,
    f,
    { delay = 0, done = delay || delay === 0 } = {}
  ) =>
    jest.fn((...args) => {
      // Register the call.
      all(type, args);

      // Define the result.
      let res;
      if (!delay && delay !== 0) {
        res = typeof f === 'function' ? f(...args) : f;
      } else {
        res = wait(delay).then(
          () => (typeof f === 'function' ? f(...args) : f)
        );
      }

      // Register the result of the promise.
      if (done) {
        res = res.then(
          res_ => {
            all(`${type} resolution`, res_);
            return res_;
          },
          err => {
            all(`${type} rejection`, err);
            throw err;
          }
        );
      }
      return res;
    });
  return { all, spy: createSpy };
};

afterEach(() => {
  jest.resetAllMocks();
});

describe('runExperiment', () => {
  it('runs as expected if every app handlers are provided', async () => {
    const spyGroup = createDelayedSpiesGroup();

    runTrials.mockImplementation(spyGroup.spy('runTrials'));

    // Connection fake implementation.
    const connection = {
      mockName: 'connection',
      // app.start and get run are supposed to be called in parallel. The delay
      // forces be resolved after app.start.
      getRun: spyGroup.spy('connection.getRun', { id: 'runId' }, { delay: 2 }),
      disconnect: spyGroup.spy('connection.disconnect')
    };
    // App fake implementation.
    const app = {
      mockName: 'app',
      start: spyGroup.spy('app.start'),
      initBlock: spyGroup.spy('app.initBlock'),
      runTrial: spyGroup.spy('app.runTrial', trial => ({
        result: `${trial.block.number}-${trial.number}`
      })),
      end: spyGroup.spy('app.end', null, { delay: false })
    };
    // Storage fake implementation
    const runStorage = {
      mockName: 'runStorage',
      set: spyGroup.spy('storage.set', null, { delay: false }),
      remove: spyGroup.spy('storage.remove', null, { delay: false })
    };
    await runExperiment(app, {
      experimentId: 'xpId',
      runId: 'runId',
      connection,
      runStorage
    });
    expect(spyGroup.all).toMatchSnapshot();
  });

  it('works properly even if only `app.runTrial` is provided', async () => {
    const spyGroup = createDelayedSpiesGroup();

    runTrials.mockImplementation(spyGroup.spy('runTrials'));

    // Connection fake implementation.
    const connection = {
      mockName: 'connection',
      getRun: spyGroup.spy('connection.getRun', { id: 'runId' }),
      disconnect: spyGroup.spy('connection.disconnect')
    };

    // App fake implementation.
    const app = {
      mockName: 'app',
      runTrial: spyGroup.spy('app.runTrial', trial => ({
        result: `${trial.block.number}-${trial.number}`
      }))
    };

    // Storage fake implementation
    const runStorage = {
      mockName: 'runStorage',
      set: spyGroup.spy('storage.set', null, { delay: null }),
      remove: spyGroup.spy('storage.remove', null, { delay: null })
    };

    await runExperiment(app, {
      experimentId: 'xpId',
      runId: 'runId',
      connection,
      runStorage
    });

    expect(spyGroup.all).toMatchSnapshot();
  });

  it('calls `app.crash` then throws if runTrials goes wrong', async () => {
    const spyGroup = createDelayedSpiesGroup();

    runTrials.mockImplementation(
      spyGroup.spy('runTrials', () =>
        Promise.reject(new Error('runTrials error'))
      )
    );

    // Connection fake implementation.
    const connection = {
      mockName: 'connection',
      getRun: spyGroup.spy('connection.getRun', { id: 'runId' }),
      disconnect: spyGroup.spy('connection.disconnect')
    };

    // App fake implementation.
    const app = {
      mockName: 'app',
      runTrial: spyGroup.spy('app.runTrial', trial => ({
        result: `${trial.block.number}-${trial.number}`
      }))
    };

    // Storage fake implementation
    const runStorage = {
      mockName: 'runStorage',
      set: spyGroup.spy('storage.set', null, { delay: null }),
      remove: spyGroup.spy('storage.remove', null, { delay: null })
    };

    // Run the experiment.
    await expect(
      runExperiment(app, {
        experimentId: 'xpId',
        runId: 'runId',
        connection,
        runStorage
      })
    ).rejects.toThrow('runTrials error');

    expect(spyGroup.all).toMatchSnapshot();
  });

  it('calls app.crash then throws for any possible things that can go wrong.');

  it('creates a RunInterface using stored run id if none are provided');
});
