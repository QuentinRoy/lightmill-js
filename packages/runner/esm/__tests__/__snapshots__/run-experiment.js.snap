// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`runExperiment calls \`app.crash\` then throws if runTrials goes wrong 1`] = `
[MockFunction spyGroup#all] {
  "calls": Array [
    Array [
      "connection.getRun",
      Array [],
    ],
    Array [
      "connection.getRun resolution",
      Object {
        "id": "runId",
      },
    ],
    Array [
      "storage.set",
      Array [
        "runId",
      ],
    ],
    Array [
      "runTrials",
      Array [
        Object {
          "disconnect": [MockFunction],
          "getRun": [MockFunction] {
            "calls": Array [
              Array [],
            ],
            "results": Array [
              Object {
                "isThrow": false,
                "value": Promise {},
              },
            ],
          },
          "mockName": "connection",
        },
        Object {
          "mockName": "app",
          "runTrial": [MockFunction],
        },
        1,
      ],
    ],
    Array [
      "runTrials rejection",
      [Error: runTrials error],
    ],
  ],
  "results": Array [
    Object {
      "isThrow": false,
      "value": undefined,
    },
    Object {
      "isThrow": false,
      "value": undefined,
    },
    Object {
      "isThrow": false,
      "value": undefined,
    },
    Object {
      "isThrow": false,
      "value": undefined,
    },
    Object {
      "isThrow": false,
      "value": undefined,
    },
  ],
}
`;

exports[`runExperiment runs as expected if every app handlers are provided 1`] = `
[MockFunction spyGroup#all] {
  "calls": Array [
    Array [
      "app.start",
      Array [],
    ],
    Array [
      "connection.getRun",
      Array [],
    ],
    Array [
      "app.start resolution",
      undefined,
    ],
    Array [
      "connection.getRun resolution",
      Object {
        "id": "runId",
      },
    ],
    Array [
      "storage.set",
      Array [
        "runId",
      ],
    ],
    Array [
      "runTrials",
      Array [
        Object {
          "disconnect": [MockFunction] {
            "calls": Array [
              Array [],
            ],
            "results": Array [
              Object {
                "isThrow": false,
                "value": Promise {},
              },
            ],
          },
          "getRun": [MockFunction] {
            "calls": Array [
              Array [],
            ],
            "results": Array [
              Object {
                "isThrow": false,
                "value": Promise {},
              },
            ],
          },
          "mockName": "connection",
        },
        Object {
          "end": [MockFunction] {
            "calls": Array [
              Array [],
            ],
            "results": Array [
              Object {
                "isThrow": false,
                "value": null,
              },
            ],
          },
          "initBlock": [MockFunction],
          "mockName": "app",
          "runTrial": [MockFunction],
          "start": [MockFunction] {
            "calls": Array [
              Array [],
            ],
            "results": Array [
              Object {
                "isThrow": false,
                "value": Promise {},
              },
            ],
          },
        },
        1,
      ],
    ],
    Array [
      "runTrials resolution",
      undefined,
    ],
    Array [
      "storage.remove",
      Array [],
    ],
    Array [
      "connection.disconnect",
      Array [],
    ],
    Array [
      "connection.disconnect resolution",
      undefined,
    ],
    Array [
      "app.end",
      Array [],
    ],
  ],
  "results": Array [
    Object {
      "isThrow": false,
      "value": undefined,
    },
    Object {
      "isThrow": false,
      "value": undefined,
    },
    Object {
      "isThrow": false,
      "value": undefined,
    },
    Object {
      "isThrow": false,
      "value": undefined,
    },
    Object {
      "isThrow": false,
      "value": undefined,
    },
    Object {
      "isThrow": false,
      "value": undefined,
    },
    Object {
      "isThrow": false,
      "value": undefined,
    },
    Object {
      "isThrow": false,
      "value": undefined,
    },
    Object {
      "isThrow": false,
      "value": undefined,
    },
    Object {
      "isThrow": false,
      "value": undefined,
    },
    Object {
      "isThrow": false,
      "value": undefined,
    },
  ],
}
`;

exports[`runExperiment works properly even if only \`app.runTrial\` is provided 1`] = `
[MockFunction spyGroup#all] {
  "calls": Array [
    Array [
      "connection.getRun",
      Array [],
    ],
    Array [
      "connection.getRun resolution",
      Object {
        "id": "runId",
      },
    ],
    Array [
      "storage.set",
      Array [
        "runId",
      ],
    ],
    Array [
      "runTrials",
      Array [
        Object {
          "disconnect": [MockFunction] {
            "calls": Array [
              Array [],
            ],
            "results": Array [
              Object {
                "isThrow": false,
                "value": Promise {},
              },
            ],
          },
          "getRun": [MockFunction] {
            "calls": Array [
              Array [],
            ],
            "results": Array [
              Object {
                "isThrow": false,
                "value": Promise {},
              },
            ],
          },
          "mockName": "connection",
        },
        Object {
          "mockName": "app",
          "runTrial": [MockFunction],
        },
        1,
      ],
    ],
    Array [
      "runTrials resolution",
      undefined,
    ],
    Array [
      "storage.remove",
      Array [],
    ],
    Array [
      "connection.disconnect",
      Array [],
    ],
    Array [
      "connection.disconnect resolution",
      undefined,
    ],
  ],
  "results": Array [
    Object {
      "isThrow": false,
      "value": undefined,
    },
    Object {
      "isThrow": false,
      "value": undefined,
    },
    Object {
      "isThrow": false,
      "value": undefined,
    },
    Object {
      "isThrow": false,
      "value": undefined,
    },
    Object {
      "isThrow": false,
      "value": undefined,
    },
    Object {
      "isThrow": false,
      "value": undefined,
    },
    Object {
      "isThrow": false,
      "value": undefined,
    },
    Object {
      "isThrow": false,
      "value": undefined,
    },
  ],
}
`;
